[AArch64][SME] Rewrite __arm_get_current_vg to preserve required registers

kmclaughlin-arm opened the pull request:
The documentation for the __arm_get_current_vg support routine specifies
that the following registers are call-preserved:
 - X1-X15, X19-X29 and SP
 - Z0-Z31
 - P0-P15

This patch rewrites the implementation of this routine in compiler-rt, as
the current version does not guarantee that these registers will be preserved.

kmclaughlin-arm committed:
[AArch64][SME] Rewrite __arm_get_current_vg to preserve required registers

The documentation for the __arm_get_current_vg support routine specifies
that the following registers are call-preserved:
 - X1-X15, X19-X29 and SP
 - Z0-Z31
 - P0-P15

This patch rewrites the implementation of this routine in compiler-rt, as
the current version does not guarantee that these registers will be preserved.

See https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#arm-get-current-vg

sdesmalen-arm commented on compiler-rt/lib/builtins/aarch64/sme-abi.S:
```diff
@@ -180,6 +184,48 @@ DEFINE_COMPILERRT_OUTLINE_FUNCTION_UNMANGLED(__arm_za_disable)
+DEFINE_COMPILERRT_OUTLINE_FUNCTION_UNMANGLED(__arm_get_current_vg)
+  .variant_pcs __arm_get_current_vg
+  BTI_C
+  mov x0, xzr
```
Is this necessary? I'd expect it to be overwritten by the call to `__arm_sme_state`.

kmclaughlin-arm replied:
It does get overwritten, I've removed this

sdesmalen-arm commented on compiler-rt/lib/builtins/aarch64/sme-abi-vg.c:
```diff
@@ -28,22 +28,3 @@ __attribute__((constructor(90))) static void get_aarch64_cpu_features(void) {
-
-__attribute__((target("sve"))) long
-__arm_get_current_vg(void) __arm_streaming_compatible {
-  struct SME_STATE State = __arm_sme_state();
```
Can the definition of this struct and the prototype of `__arm_sme_state` be removed as well?

kmclaughlin-arm replied:
They can indeed :)

sdesmalen-arm commented on compiler-rt/lib/builtins/aarch64/sme-abi.S:
```diff
@@ -180,6 +184,48 @@ DEFINE_COMPILERRT_OUTLINE_FUNCTION_UNMANGLED(__arm_za_disable)
+  .cfi_offset w30, -8
+  .cfi_offset w29, -16
+  mov     x17, x1
+  bl      __arm_sme_state
```
Should this call be moved into the `0:` branch?

kmclaughlin-arm committed:
- Removed extra mov at beginning of __arm_get_current_vg & moved
  __arm_sme_state call to the '0' branch.
- Removed SME_STATE struct from sme-abi-vg.c

kmclaughlin-arm committed:
- Removed relocation specifiers from CPU_FEATS_SYMBOL to be consistent with TPIDR2_SYMBOL

AtariDreams commented:
Should this be backpittef to 19.x

kmclaughlin-arm commented:
> Should this be backported to 19.x

Hi @AtariDreams, yes it should be. I will create an issue to backport it.

kmclaughlin-arm commented:
/cherry-pick 6da6772bf0a33131aa8540c9d4f60d5db75c32b5
